### PR TITLE
dhcpv6: T3771: Installation of routes for delegated prefixes, add excluded-prefix to PD

### PR DESCRIPTION
**Pull request by sarthurdev**
## Change Summary

* Install routes for delegated prefixes
* Add support for `excluded-prefix` to PD (RFC 6603)
* Fix file header, update XML include help text

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T3771

## Component(s) name

dhcpv6-server

## Proposed changes

* On PD allocation it will now install a route for that prefix to the requester's link-local address
* Add support for `excluded-prefix` to prefix delegation (RFC 6603)
* Fix header on DHCP hook script (We aren't using the credited code anymore)
* Change help text on `listen-interface` include to a generic description

## Smoketest result

```
DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_service_dhcpv6-server.py
DEBUG - test_global_nameserver (__main__.TestServiceDHCPv6Server.test_global_nameserver) ... ok
DEBUG - test_prefix_delegation (__main__.TestServiceDHCPv6Server.test_prefix_delegation) ... ok
DEBUG - test_single_pool (__main__.TestServiceDHCPv6Server.test_single_pool) ... ok
DEBUG -
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 3 tests in 11.589s
DEBUG -
DEBUG - OK
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sever-sever** [comment]
I think it is common for prefix-delegation and should be enabled by default.
 As there is no any sense in doing it manually 🤷‍♀️

**Comment by sarthurdev** [comment]
> I think it is common for prefix-delegation and should be enabled by default. As there is no any sense in doing it manually 🤷‍♀️

If there's a concensus I'll happily set it by default.

Initially thought not to change current behaviour and install routes without the user explicitly choosing to. Curious why Kea itself doesn't have a mechanism to install routes without requiring a custom script.

**Comment by sever-sever** [comment]
> > I think it is common for prefix-delegation and should be enabled by default. As there is no any sense in doing it manually 🤷‍♀️
> 
> If there's a concensus I'll happily set it by default.
> 
> Initially thought not to change current behaviour and install routes without the user explicitly choosing to. Curious why Kea itself doesn't have a mechanism to install routes without requiring a custom script.

Accel-ppp (service pppoe|ipoe server) does this by default without additional scripts/options. Having delegated prefixes without routes to them is useless in my mind.

**Comment by sarthurdev** [comment]
@sever-sever Removed CLI

**Comment by cbuechler** [comment]
> Accel-ppp (service pppoe|ipoe server) does this by default without additional scripts/options. Having delegated prefixes without routes to them is useless in my mind.

That can be a diff scenario though. You know with certainty the client is directly-reachable with accel-ppp's use cases. That's not necessarily true of the DHCPv6 server, as requests can be relayed to it. Enabling route addition by default is good IMO, as most VyOS use cases will need it. But the ability to disable it could be required in edge cases where the DHCPv6 server isn't the router for the networks, and something else is taking care of routing configuration. Adding routes in those cases on the DHCPv6 server is useless at best, and could break things. 

I'd recommend enabling route addition by default, and having a config option to disable.

**Comment by sarthurdev** [comment]
Agreed and good points, updated PR to add `disable-route-autoinstall` node.

**Comment by cbuechler** [comment]
@sarthurdev I submitted a [PR to your fork here](https://github.com/sarthurdev/vyos-1x/pull/6) with suggested changes. 

With my changes, I only see one problem. When you reboot, there are no routes added until the clients renew their DHCPv6 lease. That could be quite some time with broken v6 routing. 

The leases file doesn't contain the necessary information to add routes. My suggestion would be to either: 
1) Call the VyOS API and do `[set|del] protocols static route6 ...` so it's in the config file, rather than calling `ip -6 r` directly. 
2) Store PD+prefix+ifname+LL address in a new file under `/config/dhcp/` and populate the IPv6 routes from that during service startup or at boot time.

**Comment by sever-sever** [comment]
> @sarthurdev I submitted a [PR to your fork here](https://github.com/sarthurdev/vyos-1x/pull/6) with suggested changes.
> 
> With my changes, I only see one problem. When you reboot, there are no routes added until the clients renew their DHCPv6 lease. That could be quite some time with broken v6 routing.
> 
> The leases file doesn't contain the necessary information to add routes. My suggestion would be to either:
> 
> 1. Call the VyOS API and do `[set|del] protocols static route6 ...` so it's in the config file, rather than calling `ip -6 r` directly.
> 2. Store PD+prefix+ifname+LL address in a new file under `/config/dhcp/` and populate the IPv6 routes from that during service startup or at boot time.

Prefix delegation is not persist 
I’m as a client can get different prefixes from my ISP with any new address I get 
It is not fact if my ISP reboot I get the same address

It should be a separate feature request if you reboot often and really want it.

There should be general PD implementation in this PR

**Comment by cbuechler** [comment]
> Prefix delegation is not persist

It's persistent for the lifetime of your lease, which can be days. 

> I’m as a client can get different prefixes from my ISP with any new address I get
> It is not fact if my ISP reboot I get the same address

It's not a matter of whether you get the same address. You lose the ability to use addresses for which you have a valid lease, for up to ~half the lease length. The problem here is clients aren't renewing just because their ISP's router rebooted. The routers behind the VyOS router will be stuck with broken IPv6 routing for potentially days any time there's a reboot. 

This as it stands works sufficiently for my use case. I'm just using very short (10 minute) lease times, which makes this self-resolve in no more than ~5 minutes on reboot, but that's not advisable in a real ISP network.

**Review by sever-sever — APPROVED** [review]

**Review by c-po — APPROVED** [review]

File: src/system/on-dhcp-event.sh [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
-
-# This script came from ubnt.com forum user "bradd" in the following post
-# http://community.ubnt.com/t5/EdgeMAX/Automatic-DNS-resolution-of-DHCP-client-names/td-p/651311
-# It has been modified by Ubiquiti to update the /etc/host file
-# instead of adding to the CLI.
-# Thanks to forum user "itsmarcos" for bug fix & improvements
-# Thanks to forum user "ruudboon" for multiple domain fix
-# Thanks to forum user "chibby85" for expire patch and static-mapping
+#
+# Copyright (C) 2024 VyOS maintainers and contributors
```
Comment (farcaller) [inline src/system/on-dhcp-event.sh:3]: just a nit, but I don't think it is technically legal to take a contribution of one explicitly credited person and replace it with your own copyright unless they explicitly signed it off while copying their work verbatim.

Comment (sarthurdev) [reply src/system/on-dhcp-event.sh:3]: The source code from that script doesn't exist here anymore.

**Pull request merged**